### PR TITLE
[Pytorch lite predictor] Use KinetoEdgeCPUProfiler for operator profiling.

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
This diff changes the way operator profiling is done in lite predictor
benchmarking binary.
Instead of using custom callbacks it uses KinetoEdgeCPUProfiler to profile
events and then generate operator level metric from it.
Since KinetoEvents do not contain cpu clock time, now we report only wallclock
time.
This unifies various profiling effort that we have for benchmarking purpose. In
production we will still use observer based mechanism, but the advantage of
using kineto profiler is that we get few other things for free, such as:
- chrome trace generation.
- operator level memory profiling (to be added)
- flop counts (to be added)

Furthermore possible we can use python post processing script to parse chrome
trace and generate output similar to torch.profiler. (To be done)

Test Plan:
aibench run
Model without debug info:
https://www.internalfb.com/intern/aibench/details/219598441154763
Model with debug info and `--print_module_info true` (see Operator summary has now module hierarchy information).
https://www.internalfb.com/intern/aibench/details/617154236292985

Differential Revision: D30327514

### CODE REVIEW

File: torch/csrc/jit/mobile/interpreter.cpp [hunk 1]
```diff
@@ -57,6 +57,11 @@ bool InterpreterState::run(Stack& stack) {
       auto inst_with_handle = code_->instructions_with_handles_.at(pc);
       Instruction inst = inst_with_handle.instruction;
       DebugHandle debug_handle = inst_with_handle.debug_handle;
+      // If no valid debug handle found then just log pc.
+      // This is possible when we did not save debug handles
+      // In that case post processing callback in EdgeProfile will fail
+      // to retrive usefule information.
```
Comment (raziel) [inline torch/csrc/jit/mobile/interpreter.cpp:63]:
```suggestion
      // to retrieve useful information.
```

File: torch/csrc/jit/mobile/interpreter.cpp [hunk 2]
```diff
@@ -57,6 +57,11 @@ bool InterpreterState::run(Stack& stack) {
       auto inst_with_handle = code_->instructions_with_handles_.at(pc);
       Instruction inst = inst_with_handle.instruction;
       DebugHandle debug_handle = inst_with_handle.debug_handle;
+      // If no valid debug handle found then just log pc.
+      // This is possible when we did not save debug handles
```
Comment (raziel) [inline torch/csrc/jit/mobile/interpreter.cpp:61]: This comment is a bit confusing.
It makes it sound like it was possible to do it but not anymore.
The comment also sounds like we're logging pc but that things will fail?

If logging pc's is going to cause some failure, and/or not very useful, should we instead disable logging altogether?

File: torch/csrc/jit/mobile/module.cpp [hunk 3]
```diff
@@ -145,8 +145,7 @@ std::string Module::getCallStack(const int64_t debug_handle) const {
 // We really need to change this part, so in the next step for profiling support
 // for delegates, the first thing will be to rewrite how profiling is done
 // for lite interpreter.
-std::string Module::get_forward_method_debug_info(size_t pc) const {
-  auto debug_handle = find_method("forward")->get_debug_handle(pc);
+std::string Module::get_forward_method_debug_info(int64_t debug_handle) const {
```
Comment (raziel) [inline torch/csrc/jit/mobile/module.cpp:148]: So, if the feature is enabled but the model doesn't contain debug_handle then we return the "Debug info for handle..." right?

File: torch/csrc/jit/mobile/debug_info.cpp [hunk 4]
```diff
@@ -156,8 +156,8 @@ std::string MobileDebugTable::getModuleHierarchyInfo(
     const std::string& top_module_type_name) const {
   const auto it = callstack_ptr_map_.find(debug_handle);
   if (it == callstack_ptr_map_.end()) {
-    return "Module info for handle, " + std::to_string(debug_handle) +
-        ", not found.";
+    return "Debug info for handles: " + std::to_string(debug_handle) +
```
Comment (raziel) [inline torch/csrc/jit/mobile/debug_info.cpp:159]: Should these be singular ("handle")?

File: torch/csrc/jit/mobile/profiler_edge.cpp [hunk 5]
```diff
@@ -55,8 +54,33 @@ KinetoEdgeCPUProfiler::KinetoEdgeCPUProfiler(
   trace_file_name_ = fname;
 }
 
+const std::unique_ptr<profiler::ProfilerResult>& KinetoEdgeCPUProfiler::
+    disableProfiler() {
+  TORCH_CHECK(
+      !profiler_result_,
+      "KinetoEdgeCPUProfiler already disabled. "
+      "To get list of events us getProfilerResults()");
```
Comment (raziel) [inline torch/csrc/jit/mobile/profiler_edge.cpp:62]:
```suggestion
      "To get list of events use getProfilerResults()");
```

File: torch/csrc/jit/mobile/profiler_edge.cpp [hunk 6]
```diff
@@ -55,8 +54,33 @@ KinetoEdgeCPUProfiler::KinetoEdgeCPUProfiler(
...
+  profiler_result_ = profiler::disableProfiler();
+  return profiler_result_;
+}
+
+const std::unique_ptr<profiler::ProfilerResult>& KinetoEdgeCPUProfiler::
+    getProfilerResult() {
+  TORCH_CHECK(
+      profiler_result_,
+      "KinetoEdgeCPUProfiler has not been disabled. "
+      "use disableProfiler() API instead. It returns ProfilerResult");
```
Comment (raziel) [inline torch/csrc/jit/mobile/profiler_edge.cpp:72]:
```suggestion
      "use disableProfiler() API first, which returns the ProfilerResult.");
```

**Review by raziel — COMMENTED** [review]
Thanks.
A few comments.
Mainly trying to understand the behaviour where the feature is enabled (i.e. SYMBOLICATE_MOBILE_DEBUG_HANDLE is defined), but there's no debug handles present.

File: torch/csrc/jit/mobile/interpreter.cpp (hunk 2 above)
Comment (kimishpatel) [reply torch/csrc/jit/mobile/interpreter.cpp:61]: I should remove this part "      // In that case post processing callback in EdgeProfile will fail
      // to retrive usefule information.
"
Since this PR adds `hasDebugHandles` API, if debug handles were present, then what is logged in events is debug handle, otherwise it is pc.

Now why this diverging behavior:
This is mainly to provide some unique identifier to the events, in the absence of debug_handles. Otherwise, for the purpose of generating summary from events that we use in benchmarking, two different conv ops will be lumped together as same event.

This behavior is possible because saving debug info is still optional. Once it is not, I think we can guarantee that debug handles are always present, but again user can explicitly chose to note save those also.

File: torch/csrc/jit/mobile/module.cpp (hunk 3 above)
Comment (kimishpatel) [reply torch/csrc/jit/mobile/module.cpp:148]: Yes thats right. 

So this API assumes a valid debug handle is passed. You can potentially pass a random handle and if the model contained debug information, it will just symbolicate that random handle. Not that this behavior is invalid but just pointing it out.
The context where this becomes relevant is that events log "pc" and not debug handle. In that case if you call `get_forward_method_debug_info` with pc you may just get wrong debug information. So there is also `hasDebugHandles` API which tells you that what events logged is debug handle, somewhat indirectly, and not pc. So the way use the API then is:
1. Check `hasDebugHandles` == true
2. Use the handle in events with `get_forward_method_debug_info`.

I dont quite like the disjoint nature of the two, but the reason they exist is to provide meaningful profiling information even when debug handles are not present. And in that case we tag events with PC.

I do think however that saving debug handles always, maybe unconditionally, will be more consistent at the cost always saving debug handles. Downside is that 1) debug handles now have to part of program definition and 2) older models without debug handles dont get meaningful profiling information.

**Comment by kimishpatel** [comment]
> Mainly trying to understand the behaviour where the feature is enabled (i.e. SYMBOLICATE_MOBILE_DEBUG_HANDLE is defined), but there's no debug handles present.

So right now we have saving of debug handles and symbolication of them as two separate events. You can save debug handles but not compile lib with SYMBOLICATE_MOBILE_DEBUG_HANDLE. And vice-versa. For former it will just result in empty string. For latter we will get "module info with debug handles ... not found"

File: torch/csrc/jit/mobile/interpreter.cpp (hunk 1 above)
Comment (kimishpatel) [reply torch/csrc/jit/mobile/interpreter.cpp:63]: Funny! I am compensating for lost e elsewhere. lol.

File: torch/csrc/jit/mobile/module.cpp (hunk 3 above)
Comment (raziel) [reply torch/csrc/jit/mobile/module.cpp:148]: yeah, I think we need to evaluate the cost of having debug handles since in some cases (embedded) it may be prohibitive. Ideally, we would make it a pluggable component (I guess it kind of is with the ifdef) so it can be entirely disabled, which may not be a bad idea since doesn't it assume file system access even when you use pc for logging (which won't work on embedded)?

So, I would see just the 2 options:
* With debug handles (at build time) to get the associated profiling/debug features.
* Without debug handles and you won't get any of the features nor the overhead.

Not for this PR but what do you think? We would still have to evaluate the cost of always adding debug handles (I thought you had measured it already?).

File: torch/csrc/jit/mobile/profiler_edge.cpp [hunk 7]
```diff
@@ -27,17 +26,26 @@ KinetoEdgeCPUProfiler::KinetoEdgeCPUProfiler(
   if (with_modules || with_stack) {
     auto post_processing = [this, with_stack, with_modules](
                                std::vector<profiler::KinetoEvent>& events) {
+      std::string no_debug_info("Model was not saved with debug information");
```
Comment (raziel) [inline torch/csrc/jit/mobile/profiler_edge.cpp:29]: Yeah I think this is needed for now (until we decide to always or never use debug handles).

**Review by raziel — APPROVED** [review]
Thanks!
Just have a comment for a follow up on whether we want to make debug handles always present, or disable the feature entirely (rather than fallback to pc).

File: torch/csrc/jit/mobile/module.cpp (hunk 3 above)
Comment (kimishpatel) [reply torch/csrc/jit/mobile/module.cpp:148]: > With debug handles (at build time) to get the associated profiling/debug features.

Currently this feature is supported via 1) save debug handles and 2) build with eager symbolication.

> Without debug handles and you won't get any of the features nor the overhead.

Similarly this feature is available via 1) not save debug handles and 2) not build with symbolication source code.

I think the options that you propose make sense. My only concern is providing the support for the 2 options via single switch. Given that model training/export pipeline is quite separate from deployment, we will continue to have to provide two switches, 1 and 2 above.

> We would still have to evaluate the cost of always adding debug handles (I thought you had measured it already?).

Yes we did that internally in terms of its impact on the model size, which corresponds to 1 above. For 2 we have not evaluated the impact on code size yet.

> Then, for older models we would probably just not log anything.

Right. So this would just look like how current PR responds to profiling when `hasDebugHandle` returns false, right?

File: torch/csrc/jit/mobile/profiler_edge.cpp (hunk 7 above)
Comment (kimishpatel) [reply torch/csrc/jit/mobile/profiler_edge.cpp:29]: I think we would need this, as you pointed out, for embedded or other deployment scenario where we explicitly dont want any overhead.
